### DISCUSSION TITLE
Pipelining

### DISCUSSION THREAD

**Discussion by petru-motrescu** [redis/lettuce#2259] [Q&A] [answered]
Hi. My team has taken Lettuce into use in a Kotlin grpc service. We have a use case where we need to increment and expire multiple counters in the span of a single grpc request. We'd like to pipeline all these commands and send them to the server in 1 go. What is the recommended way to achieve this with Lettuce? At the moment we share a single instance of `StatefulRedisConnection` across all threads. Would we need to create multiple of these, one for each thread?

Thank you 🙇

**Comment by mp911de** [comment, marked as answer]
Lettuce uses pipelining as the general mode of operations, the client doesn't await the completion of a command before it sends the next one. 

If you use the async API, then you can ensure that commands are issued immediately and serially without awaiting continuations/subscriptions/backpressure demand (in contrast to the Coroutines/reactive API).

A single connection will give you the best performance. There are other options (e.g. manual command flushing) but that approach requires a connection per process which seems to be more expensive than a single-connection approach.
